operator/pkg/ciliumendpointbatch/reconciler.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ciliumendpointbatch
+
+import (
+	"context"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	cilium_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	clientset "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/typed/cilium.io/v2alpha1"
+	"github.com/sirupsen/logrus"
+)
+
+// reconciler is used to sync the current (i.e. desired) state of the CEBs in datastore into current state CEBs in the k8s-apiserver.
+// The source of truth is in local datastore, client reconcile with apiserver at regular intervals.
+type reconciler struct {
+	client     clientset.CiliumV2alpha1Interface
+	cebManager cebManager
+}
+
+// newReconciler creates and initializes a new reconciler.
+func newReconciler(client clientset.CiliumV2alpha1Interface, cebMgr cebManager) *reconciler {
+	return &reconciler{
+		client:     client,
+		cebManager: cebMgr,
+	}
+}
+
+// reconcileWithServer gets the list of CEBs to create, update and delete and syncs with the api-server.
+// On error, it returns the list of failed createCebs, updateCebs and deleteCebs to the caller.
+func (r *reconciler) reconcileWithServer(cebToCreate, cebToUpdate, cebToDelete string) error {
+
+	// Update the CEB
+	if cebToUpdate != "" {
+		var err error
+		var retCeb, ceb *cilium_v2.CiliumEndpointBatch
+		// Before syncing with ApiServer, get list of removed CEPs
+		remCeps := r.cebManager.getRemovedCeps(cebToUpdate)
+		// Get the copy of Ceb from the cebManager
+		if ceb, err = r.cebManager.getCebFromCache(cebToUpdate); err != nil {
+			return err
+		}
+
+		if retCeb, err = r.client.CiliumEndpointBatches().Update(
+			context.TODO(), ceb, metav1.UpdateOptions{}); err != nil {
+			log.WithError(err).WithFields(logrus.Fields{
+				"ceb-name": cebToUpdate,
+			}).Info("Error Updating CiliumEndpointBatch")
+			return err
+		}
+		// Update local datastore CEB with latest object metadata values.
+		r.cebManager.updateCebInCache(retCeb, false)
+		// Since local copy of CEB is synced with api-server, we don't need to track anymore removed
+		// CEPS.
+		r.cebManager.clearRemovedCeps(cebToUpdate, remCeps)
```

Comment:
why do we need to pass in `remCeps`? Isn't that already encoded in `cebToUpdate.removedCeps`?

Reply:
The whole point of passing `remCeps` is `not` to clear all `removedCeps`.  The CEB we are updating might have removed only 1 `cep`, let's call it `cep-abcd`. During CEB update,  maybe one more CEP might have removed from CEB, let's call it `cep-ab12`. 
So now we want to remove only `cep-abcd` in removed list, because api-server also sees only `cep-abcd` is removed from CEB.

Reply:
If one of the `remCeps` is re-created, 
1)  In `InsertCepInCache` recreated cep is added back to `ceb.Endpoints` and re-created CEP will be removed from `removedCeps`.

Reply:
> re-created CEP will be removed from removedCeps.

But it is still present in the `remCeps` variable, so whe `clearRemovedCeps` is called, the CEP will be removed from the cache.

Reply:
As i told earlier, if cep is re-created, that `cep` is removed from `ceb.removedCeps` map.
If `clearRemovedCeps` called with same ` re-created cep name`, current code do nothing.
```
	// Delete removed CEPs from caches.
	for cn := range remCeps {
		if _, ok = ceb.removedCeps[cn]; ok {
			c.cacheCepMap.deleteCep(cn)
			delete(ceb.removedCeps, cn)
		}
	}
```

Reply:
I see. Thanks for explanation.